src/main/kotlin/com/amazon/opendistroforelasticsearch/indexstatemanagement/ManagedIndexRunner.kt

```diff
@@ -47,60 +67,141 @@ object ManagedIndexRunner : ScheduledJobRunner {
+        // TODO: Evaluate which state, action, step we're on from the policy and ManagedIndexMetaData
+
+        // TODO: Execute the step
+
+        // TODO: Update the cluster state with the updated ManagedIndexMetaData
+    }
 
-        if (existingManagedMetadata != null) {
-            logger.info("Start from where we left off. $existingManagedMetadata")
+    private suspend fun initManagedIndex(
+        managedIndexConfig: ManagedIndexConfig,
+        managedIndexMetaData: ManagedIndexMetaData?
+    ) {
+        // If policy does not currently exist, get policy by name and save it to ManagedIndexConfig
+        var policy: Policy? = managedIndexConfig.policy
+        if (policy == null) {
+            policy = getPolicy(managedIndexConfig.policyName)
+
+            if (policy != null) {
+                val saved = savePolicyToManagedIndexConfig(managedIndexConfig, policy)
+                if (!saved) return // If we failed to save the policy, don't initialize ManagedIndexMetaData
+            }
+            // If policy is null we handle it in the managedIndexMetaData by moving to ERROR
+        }
+
+        // Initialize ManagedIndexMetaData
+        if (managedIndexMetaData == null) {
+            initializeManagedIndexMetaData(managedIndexConfig, policy)
         } else {
-            logger.info("Start from default state.")
+            // TODO: This could happen when deleting a ManagedIndexConfig document
+            // We should compare the cached policy with the ManagedIndexMetaData and make sure we have the same policy, seqNo, primaryTerm
+            // otherwise update to ERROR with policy/version conflict
+        }
+    }
+
+    private suspend fun getPolicy(policyName: String): Policy? {
+        val getRequest = GetRequest(INDEX_STATE_MANAGEMENT_INDEX, policyName).preference("_local")
+        val getResponse: GetResponse = client.suspendUntil { get(getRequest, it) }
+        if (!getResponse.isExists || getResponse.isSourceEmpty) {
+            return null
         }
 
-        // Update ManagedIndexMetadata
-        val updateMetadata = ManagedIndexMetaData(
-            indexMetaData.index.name,
-            indexMetaData.index.uuid,
-            "${indexMetaData.index.name}_POLICY_NAME",
-            "${indexMetaData.index.name}_POLICY_VERSION",
-            "${indexMetaData.index.name}_STATE",
-            "${indexMetaData.index.name}_STATE_START_TIME",
-            "${indexMetaData.index.name}_ACTION_INDEX",
-            "${indexMetaData.index.name}_ACTION",
-            "${indexMetaData.index.name}_ACTION_START_TIME",
-            "${indexMetaData.index.name}_STEP",
-            "${indexMetaData.index.name}_STEP_START_TIME",
-            "${indexMetaData.index.name}_FAILED_STEP"
-        )
+        val policySource = getResponse.sourceAsBytesRef
+        // Intellij complains about createParser/parseWithType blocking because it sees they throw IOExceptions
+        return withContext(Dispatchers.IO) {
```

Comment:
still using IO Dispatcher?

Reply:
This is because intellij is complaining that we are blocking here since the parser/xcp.nextToken/etc. says it throws an IOException

Vinoo looked into this before for alerting and said this was just a workaround to get rid of the issue:
https://github.com/opendistro-for-elasticsearch/alerting/blob/master/alerting/src/main/kotlin/com/amazon/opendistroforelasticsearch/alerting/MonitorRunner.kt#L455